Fixed Some Bugs in '1.sh' + Some Tidying

terminalforlife opened the pull request:
I'll leave comments in various places, as there are a few things to address here.

Despite these improvements and fixes, I still think you'll run into issues. Most especially because of your line 20 `if` condition; it looks like you're using an integer comparison operator on a string comparison, which will result in an error.

Commit:
Updated 1.sh; fixed bugs; tidied a bit

terminalforlife commented on 1.sh:
```diff
@@ -1,35 +1,34 @@
 #! /bin/sh
 
-if [-d $1] && [-d $2]
+if [ -d "$1" ] && [ -d "$2" ]
```
In shell, spacing is _very_ important. Since `[` is essentially a command (a shell builtin), the spacing rules must apply.

Remember to protect your variables (even your assignments, where necessary) to ensure bugs don't arise in your programs. It's very important. Had I not fixed this, you'd have ran into errors on this line alone, due to the possibility of passing one or two directory names with spaces, or perhaps special characters bash would otherwise interpret.

TheScarletArrow replied:
Thanks for your contribution. It seems my common mistake was with "$x", and I aprriciate your correction! I will upload newer version soon, hope you can check it and spot the mistakes.

terminalforlife commented on 1.sh:
```diff
@@ -1,35 +1,34 @@
+elif [ ! -d "$2" ] 
 	echo "Flash path doesn't exist"
-	mkdir $2
+	mkdir "$2"
```
This is another example of spacing being important. If your first argument for the script were a directory named `My Dir` then this command here would have created two directories, `My` and `Dir`.

terminalforlife commented on 1.sh:
```diff
@@ -1,35 +1,34 @@
 fi
 
-for file in $1/*
+for file in "$1"/*
```
Notice how I didn't wrap `/*` in quotes, too? That's because bash would then not have been able to interpret the asterix (glob) for Filename Expansion. Technically, it could be written as `"$1/"*`, but I found the other approach more readable.

terminalforlife commented on 1.sh:
```diff
@@ -1,35 +1,34 @@
-		echo "Copying files to $2"
-		cp $file $2
+	echo "$(basename $file)"
+	flashfile="$2/$(basename $file)"
```
I may have misunderstood what you were wanting here, but it looks as though you've accidentally added an additional layer of command substitution (nested command substitution). The variable `flashfile` will now be assigned only the contents of `$2/$(basename $file)".

To break this down:

`$2` will be expanded to the value assigned to the 2nd positional parameter given to your script; the second directory name.

The `$(basename $file)` is command substitution, in its more modern form; it used to be done with graves (```) but that had stylistic and practical issues, so `$()` came to be. You probably figured this out by now, but command substitution allows you to place the output of a command (or even entire blocks of code!) almost anywhere; it allows for a LOT of flexibility.

The `$file` variable is expanded within the command substitution to the value assigned to the variable, which is the currently iterated file in the `for` loop.

terminalforlife commented on 1.sh:
```diff
@@ -1,35 +1,34 @@
+
+	if [ ! -f "$flashfile" ]; then
+		echo "Copying files to '$2'"
+		cp "$file" "$2"
```
Again, protect your variables, and even your strings! In the case, the outcome could be quite dangerous. (data loss due to overwrites)

Here's an example situation:

I want to copy file `My First File` which is in the current directory, to the directory `/home/user/`. Your way would be as follows:

```bash
cp My First File /home/user/
```

This would, however, move files `My`, `First`, and `File` to `/home/user/`, overwriting anything with those names in the target directory. The correct way, therefore, would be, at the very least:

```bash
cp 'My First File' /home/user/
```

Since `/home/user/` has no spaces, variables, or any special characters bash would otherwise interpret, you're safe to leave it as-is. The `My First File`, however has spaces, as so needs to be protected by either single or double quotes; in this case, it doesn't matter which.

terminalforlife commented on 1.sh:
```diff
@@ -1,35 +1,34 @@
-	*) echo "Wrong input"
-	;;
-	esac
+		read -e -n 1 -p "Flash version is newer, replace existing one? [Y|N] "
```
While not necessary, I would suggest this approach as it's more concise, in my opinion.

The bash builtin `read` is capable of providing a prompt to the user, by using the `-p` flag. See `help read` for further, albeit limited information. The `-e` flag allows for a new line to be displayed, where otherwise it wouldn't. The `-n` flag is for having read accept only the given number (in this case, `1`) of characters before immediately (without the need for the Enter or Return key to be pressed) proceeding.

Lastly, here, `[Y|N]` is a traditional approach to displaying a key choice to the user, but of course is by no means required.

terminalforlife commented on 1.sh:
```diff
@@ -1,35 +1,34 @@
+				echo "Staying as we were" 
+				exit 2 ;;
+			*)
+				echo "Wrong input" ;;
```
Remember, a `;;` must be given after the code here in a case statement, as shown above. It can either be on the end of the last line of command(s) or alone on one of the following lines.

Commit:
Resolve conflict after merge with master (+more)

Regarding the "+more" part:

I may not have realised recently, when I made changes to this file, -
that you were in-fact using the Bourne Shell (`sh`) _not_ the Bourne
Again Shell (`bash`), so I've undone my changes and hopefully once
again made it compatible with the correct shell. Sorry about that.

There were some bizarre quotations when you used `echo`, so I sorted
that out, and I think it was down to you either being confused about
rules and best practices with variables, or you wanted to have the user
see the contents of the variables _quoted_, for easy distinction; with
the latter in mind, I've single-quoted them, as seems commonplace.

Incorrect (albeit technically valid _here_) quotation:

  echo "this is a "$variable" which holds stuff"

Correct quotation:

  echo "this is a \"$variable\" which holds stuff"

Also correct, if preferred:

  echo "this is a '$variable' which holds stuff"

So why was the first one incorrect, yet still displayed?

It's partly down to how echo works, but mainly down to how shell works.
You've probably already discovered that things in shell should be
space-delimited, as that's usually how they are, by default, -
interpreted, but in this case, assuming the variable isn't assigned a
value with spaces or shell-interpretable characters, it's still valid.

Ordinarily, however, this may be a risky crutch to lean on, so it's
usually best to play it safe, in this regard.

Admittedly, it's not black and white, but I imagine after a while, -
you'll see what I'm getting at. ;)

Commit:
Resolve conflict after merge (re-applied)

Must have been timing earlier, but I've applied my undoings again. lol
Again, I apologise for any inconvenience and/or confusion.

Commit:
Resolve conflict following merge with master

Commit:
Merge branch 'master' of https://github.com/TheScarletArrow/bashone
